src/main/scala/midas/models/dram/Util.scala

```diff
@@ -407,12 +408,57 @@ class AddressCollisionChecker(numReads: Int, numWrites: Int, addrWidth: Int)(imp
 }
 
 
-class HistogramReadoutIO(val binAddrBits: Int) extends Bundle {
+class CounterReadoutIO(val addrBits: Int) extends Bundle {
   val enable = Input(Bool()) // Set when the simulation memory bus whishes to read out the values
-  val addr = Input(UInt(binAddrBits.W))
+  val addr = Input(UInt(addrBits.W))
   val dataL = Output(UInt(32.W))
   val dataH = Output(UInt(32.W))
 }
+
+class CounterIncrementIO(val addrBits: Int, val dataBits: Int) extends Bundle {
+  val enable = Input(Bool())
+  val addr = Input(UInt(addrBits.W))
+  // Pass data 2 cycles after enable
+  val data = Input(UInt(dataBits.W))
+}
+
+class CounterTable(addrBits: Int, dataBits: Int) extends Module {
+  val io = IO(new Bundle {
+    val incr = new CounterIncrementIO(addrBits, dataBits)
+    val readout = new CounterReadoutIO(addrBits)
+  })
+
+  require(dataBits > 32)
+  val memDepth = 1 << addrBits
+
+  val counts = SyncReadMem(memDepth, UInt(dataBits.W))
+
+  val s0_readAddr = Mux(io.readout.enable, io.readout.addr, io.incr.addr)
+  val s1_readAddr = RegNext(s0_readAddr)
+  val s1_readData = counts.read(s0_readAddr)
+  val s1_valid = RegNext(io.incr.enable)
```

Comment:
Do none of these need to be reset?

Reply:
Hmm, yeah i see why this works; It's probably just safer to reset these though, in case someone doesn't actually tare their histogram first.